mesonbuild/compilers/c.py

```diff
@@ -38,6 +39,116 @@
 
 
 class CCompiler(Compiler):
+
+    # These functions are based on the following code:
+    # https://git.savannah.gnu.org/gitweb/?p=autoconf-archive.git;a=blob_plain;f=m4/ax_gcc_func_attribute.m4,
+    # which is licensed under the following terms:
+    #
+    #   Copyright (c) 2013 Gabriele Svelto <gabriele.svelto@gmail.com>
+    #
+    #   Copying and distribution of this file, with or without modification, are
+    #   permitted in any medium without royalty provided the copyright notice
+    #   and this notice are preserved.  This file is offered as-is, without any
+    #   warranty.
+    __FUNC_ATTRIBUTES = {
```

Comment:
`__` has a special meaning [https://docs.python.org/3/reference/lexical_analysis.html?highlight=mangle#reserved-classes-of-identifiers]. I doubt that's needed here. In fact it's probably better to just raise this to a top-level variable.

Reply:
I am aware of python's name mangling. I think this deserves to be class private, it exists to feed a single method in the class `attribute_check_function`. The only reason I put it into the class rather than into the method is to avoid re-initializing that dict on every function call.

Reply:
Class private variables aren't really used, because they are confusing, for example when debugging, and make little sense in practice. That class in unlikely to be subclassed, in which case there's no point to use a class-private variable, hidden from subclasses. But if it was ever subclassed for some reason, it's possible that the child would want to look at this dictionary, in which case is better for it be a normal field. Either way, `__` is not useful.

Reply:
I cannot imagine a case where a subclass would want to look at this dictionary. Since it is class level, if a subclass where to modify it, then it would effect the parent class and other siblings as well. If a subclass really did want to look at it, the right way would be to do something like:
```python
@classmethod
def attribute_check_function(self, attr):
    if some_condition:
         return some_value
    return super().attribute_check_function(attr)
```

Reply:
That's ... overengineering ;)

OK, let's say I want to write a helper (e.g. in meson --help) that prints a list of known attributes. Name mangling makes this much harder than necessary.

Anyway, the way I think about private attributes, is that it's a cargo cult from languages like java, to be avoided unless there's a good reason. There is no reason here.

Reply:
My experience is that if it's public, it needs to be safe to modify. The fact is it's not safe to modify, modifying it will lead to unexpected behavior that will be especially hard to debug. But, I have another idea.

Reply:
From PEP 416: "use of frozendict is low. Those that do use it tend to use it as a hint only, such as declaring global or class-level "constants": they aren't really immutable, since anyone can still assign to the name."

In fact there are mutable global variables in meson:
```
mesonbuild/modules/i18n.py:PRESET_ARGS = {
mesonbuild/modules/qt.py:_QT_DEPS_LUT = {
```

But if you really want to make it read-only, just use `types.MappingProxyType`.

Reply:
And yeah, making it a local variable seems to be the worst solution.

Reply:
in the case of the qt variable though it's marked as module protected.

Ah, I always forget about the Proxy types, meson is the only project I work on that is py3k only.